web/src/information_density.ts

```diff
@@ -171,3 +200,259 @@ export function initialize(): void {
     calculate_timestamp_widths();
     determine_container_query_support();
 }
+
+export const information_density_properties_schema = z.enum([
+    "web_font_size_px",
+    "web_line_height_percent",
+]);
+
+export function enable_or_disable_control_buttons($container: JQuery): void {
+    const info_density_properties = z
+        .array(information_density_properties_schema)
+        .parse(["web_font_size_px", "web_line_height_percent"]);
+    for (const property of info_density_properties) {
+        const $button_group = $container.find(`[data-property='${CSS.escape(property)}']`);
+        const $current_elem = $button_group.find<HTMLInputElement>(".current-value");
+        const current_value = Number.parseInt($current_elem.val()!, 10);
+
+        $button_group
+            .find(".default-button")
+            .prop("disabled", current_value === INFO_DENSITY_VALUES_DICT[property].default);
+        $button_group
+            .find(".increase-button")
+            .prop("disabled", current_value >= INFO_DENSITY_VALUES_DICT[property].maximum);
+        $button_group
+            .find(".decrease-button")
+            .prop("disabled", current_value <= INFO_DENSITY_VALUES_DICT[property].minimum);
+    }
+}
+
+const line_height_supported_values = [122, 131, 140, 149, 158];
+
+export function find_new_supported_value_for_setting(
+    $elem: JQuery,
+    property: "web_font_size_px" | "web_line_height_percent",
+    current_value: number,
+): number {
+    if (current_value > INFO_DENSITY_VALUES_DICT[property].maximum) {
+        return INFO_DENSITY_VALUES_DICT[property].maximum;
+    }
+
+    if (current_value < INFO_DENSITY_VALUES_DICT[property].minimum) {
+        return INFO_DENSITY_VALUES_DICT[property].minimum;
+    }
+
+    assert(property === "web_line_height_percent");
```

Comment:
I guess it'd be nice to have a comment here, noting that we only support integer font sizes, but there are valid line height values in between the steps here.

Reply:
``` diff
@@ -242,6 +243,10 @@ export function find_new_supported_value_for_setting(
         return INFO_DENSITY_VALUES_DICT[property].minimum;
     }
 
+    // We know the value is inside the range of valid values, but not
+    // a recommended value. This is only possible with line height,
+    // where we allow any integer in the database, but only offer
+    // certain steps in the UI.
     assert(property === "web_line_height_percent");
 
     if ($elem.hasClass("increase-button")) {
```